m4/odp_atomic.m4

```diff
@@ -126,12 +126,13 @@ AC_CACHE_CHECK([whether -latomic is needed for 64-bit atomic compare exchange],
 	       [odp_cv_atomic_needed_64bit_cmp_exc], [dnl
 AC_LINK_IFELSE(
   [AC_LANG_SOURCE([[
+    #include <stdbool.h>
     #include <stdint.h>
     static uint64_t loc;
     int main(void)
     {
         uint64_t exp = 0;
-        uint64_t = __atomic_compare_exchange_n(&loc, &exp, 1, 1,
+        bool res = __atomic_compare_exchange_8(&loc, &exp, 1, 1,
```

Comment:
Would addition of the bool variable fix the issue. Our code base uses compare_exchange_n(), but not compare_exchange_8(). So, I'd vote for the minimal change here.

bool res = __atomic_compare_exchange_n(...)

Reply:
Noticed that commit log mentions that compare_exchange_n() didn't work on some cross compile case. If this is the case, maybe we then merge this compare_exchange_8() version and change back to compare_exchange_n() if problems arise.

Reply:
Yep, it was clang that failed. Just to be sure I've added a temporary build on [my travis instance](https://travis-ci.org/github/mvl-skardach/odp/builds/703525499) changing back to _n(). Just to see if that previous failure was not some artifact.

Reply:
Still fails in clang if using _n(): https://travis-ci.org/github/mvl-skardach/odp/jobs/703525540#L9500

Reply:
I do not understand this and I guess I missed the explanation earlier today. Why _n is not enough? Using the explicit functions look a bit questionable to me. Are we supposed to even know they are there and named like that? And if the actual code is using _n, why is that not sufficient here? Could you elaborate?

Reply:
ipfragreass code assumes that when __SIZEOF_POINTER__ == 4, compare_exchange_n() will be able to handle uint64_t types. Maybe this support should be checked instead, and disable ipfragreass compilation when it is not supported.

It may be compare_exchange_n() has additional checks (64 bit atomics support on 32 bit) and it won't compile, but compare_exchange_8() don't have such checks (as the front end _n() function have already done that). So, there may be a valid reason _n() won't compile, but direct _8() does (but could fail to a race condition on run time).

Reply:
That would make sense.

Reply:
> Using the explicit functions look a bit questionable to me.

Explicit `compare_exchange_n()` is used just several lines below (line [144](https://github.com/OpenDataPlane/odp/pull/976/files#diff-363fe833c8e839044373736c07845f58R144)):

```
if test "x$odp_cv_atomic_needed_64bit_cmp_exc" = "xyes" ; then
   AC_CHECK_LIB(	   AC_CHECK_LIB(
     [atomic], [__atomic_compare_exchange_8],	     [atomic], [__atomic_compare_exchange_8],
     [m4_default([$1], [:])],	     [m4_default([$1], [:])],
     [m4_default([$2], [:])])	     [m4_default([$2], [:])])
fi
```

So question is whether that check is also questionable and if so then whole odp_atomic.m4 should be reviewed. I've started with an assumption that it's acceptable.

> ipfragreass code assumes that when SIZEOF_POINTER == 4, compare_exchange_n() will be able to handle uint64_t types. Maybe this support should be checked instead, and disable ipfragreass compilation when it is not supported.

I don't think it's a problem of `compare_exchange_n()` not handling the uint64_t because it does as soon as you add `-latomic`. The problem is with clang and `_n()` version (see the build [here](https://travis-ci.org/github/mvl-skardach/odp/jobs/703525540#L445)) is that it passes the `AC_LINK_IFELSE` [check](https://travis-ci.org/github/mvl-skardach/odp/jobs/703525540#L11714) which causes `-latomic` not to be added to `ATOMIC_LIBS_OPT`. Later on it causes the ipfragreass to fail because it's missing the `-latomic`.